### ISSUE TITLE
The generic Theory method does not support decimal argument type.

### ISSUE THREAD

**Issue by AlexMihalev** [xunit/xunit#2418]
Here test example:

```csharp
public class UnitTest1
{
    public static IEnumerable<object[]> ValuesWithDecimal() 
    {
        yield return new object[] { (decimal)42 };
    }

    public static IEnumerable<object[]> ValuesWithoutDecimal() 
    {
        yield return new object[] { (long)42 };
        yield return new object[] { (ulong)42 };
        yield return new object[] { (double)42 };
    }

    [Theory]
    [MemberData(nameof(ValuesWithDecimal))]
    public void Fails<T>(T value) => Assert.NotNull(value);

    [Theory]
    [MemberData(nameof(ValuesWithoutDecimal))]
    public void Pass<T>(T value) => Assert.NotNull(value);
}
```

And test output:
```
[xUnit.net 00:00:00.37]     xunit_bug.UnitTest1.Fails [FAIL]
  Failed xunit_bug.UnitTest1.Fails [1 ms]
  Error Message:
   System.Reflection.AmbiguousMatchException : Ambiguous match found.
  Stack Trace:
     at System.DefaultBinder.SelectMethod(BindingFlags bindingAttr, MethodBase[] match, Type[] types, ParameterModifier[] modifiers)
   at System.RuntimeType.GetMethodImplCommon(String name, Int32 genericParameterCount, BindingFlags bindingAttr, Binder binder, CallingConventions callConv, Type[] types, ParameterModifier[] modifiers)
   at System.Type.GetMethod(String name, BindingFlags bindingAttr, Binder binder, CallingConventions callConvention, Type[] types, ParameterModifier[] modifiers)
   at System.Reflection.RuntimeReflectionExtensions.GetRuntimeMethod(Type type, String name, Type[] parameters)
--- End of stack trace from previous location ---

Failed!  - Failed:     1, Passed:     3, Skipped:     0, Total:     4, Duration: 3 ms - xunit-bug.dll (net6.0)
```

After some debugging, I found that the original exception is raised from [here](https://github.com/xunit/xunit/blob/d53d0761f997ae132b9147c2347659029c9d0e23/src/xunit.v3.common/Utility/TypeUtility.cs#L184)

And looks like the root cause that TypeUtility.ResolveMethodArguments is don't expect that the passed method can be generic.

Here more complete exception trace:
```
  System.Private.CoreLib.dll!System.DefaultBinder.SelectMethod(System.Reflection.BindingFlags bindingAttr, 
  System.Reflection.MethodBase[] match, System.Type[] types, System.Reflection.ParameterModifier[] modifiers) Line 608	C#
  System.Private.CoreLib.dll!System.RuntimeType.GetMethodImplCommon(string name, int genericParameterCount, 
  System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Reflection.CallingConventions callConv, 
  System.Type[] types, System.Reflection.ParameterModifier[] modifiers) Line 2790	C#
  System.Private.CoreLib.dll!System.Type.GetMethod(string name, System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder 
  binder, System.Reflection.CallingConventions callConvention, System.Type[] types, System.Reflection.ParameterModifier[] modifiers) Line 314	C#
  System.Private.CoreLib.dll!System.Reflection.RuntimeReflectionExtensions.GetRuntimeMethod(System.Type type, string name, 
  System.Type[] parameters) Line 74	C#
  xunit.execution.dotnet.dll!Xunit.Sdk.TypeUtility.PerformDefinedConversions(object argumentValue = 42, System.Type conversionDeclaringType = {Name = "Decimal" FullName = "System.Decimal"}) Line 146	C#
generic type here --> xunit.execution.dotnet.dll!Xunit.Sdk.TypeUtility.TryConvertObject(object argumentValue = 42, System.Type parameterType = {Name = "T" FullName = null}) Line 127	C#
  xunit.execution.dotnet.dll!Xunit.Sdk.TypeUtility.ResolveMethodArguments(System.Reflection.MethodBase testMethod, object[] arguments = {object[1]}) Line 98	C#
  xunit.execution.dotnet.dll!Xunit.Sdk.TestMethodTestCase.Initialize() Line 319	C#
  xunit.execution.dotnet.dll!Xunit.Sdk.XunitTestCase.Initialize() Line 85	C#
  xunit.execution.dotnet.dll!Xunit.Sdk.TestMethodTestCase.EnsureInitialized() Line 249	C#
  xunit.execution.dotnet.dll!Xunit.Sdk.TestMethodTestCase.DisplayName.get() Line 77	C#
...
```

[Here](https://github.com/AlexMihalev/xunit-bug) sample repository.

> PS: If you are interested, I can try to fix this bug.

**Comment by SteveDunn** [comment]
I think this is covered here: https://stackoverflow.com/questions/507528/use-decimal-values-as-attribute-params-in-c

TL;DR: `decimal` isn't a primitive type and cannot be used as a constant value

**Comment by AlexMihalev** [comment]
I don't think the problem is that the **decimal** is not a primitive type. I already can use custom type in such a way without any issues. Here is the proof:
```csharp
using System.Collections.Generic;
using Xunit;

namespace xunit_bug;

public class UnitTest1
{
    public class NonPrimitiveType
    {
        public int Value { get; set; }
    }

    public static IEnumerable<object[]> Values() 
    {
        yield return new object[] { new NonPrimitiveType { Value = 42 } };
    }

    [Theory]
    [MemberData(nameof(Values))]
    public void Pass<T>(T value) => Assert.NotNull(value);
}
```
works as expected.

**Comment by bradwilson** [comment]
Definitely seems like a bug. Will definitely take a fix against both the `v2` and `main` (v3) branches. Thanks!

**Comment by louis-z** [comment]
Hey,
I'd like take a stab at it, if no one else currently is.

**Comment by louis-z** [comment]
Calls to `System.Reflection.RuntimeReflectionExtensions.GetRuntimeMethod()` assume that only one method will match the 
- method name (e.g. `op_Implicit` or `op_Explicit`), and
- parameter type (e.g. `Decimal`)  

combination. If there are more than one match - which is the case with `Decimal`'s conversion methods -, a `System.Reflection.AmbiguousMatchException` is thrown.

Just submitted a PR that proposes a fix. I also tweaked tests to reproduce those conflicting conversion operators.

**Comment by bradwilson** [comment]
Available in v2: [`2.4.2-pre.21`](https://myget.org/feed/xunit/package/nuget/xunit/2.4.2-pre.21)
Available in v3: [`0.1.1-pre.27`](https://myget.org/feed/xunit/package/nuget/xunit.v3/0.1.1-pre.27)

**Comment by jzabroski** [comment]
@bradwilson I think this may be causing a slight (and very annoying) regression in InlineData theories for the special `0` literal in C#.

```c#
[Theory]
[InlineData(null)]
[InlineData(0)] // fails at runtime because it's not "0.0"
[InlineData(1.0)]
public void TestDoubles(double? d)
{
  Assert.Equal(d, d);
}
```

The above program works without errors in 2.4.1.  On 2.4.2, it yields the following stack trace and test failure - it also seems to cause the tests to run ridiculously slow, but that could be some other error I have not seen yet, since I have at least 200 tests failing due to this issue around the `0` literal failing.


System.ArgumentException
Object of type 'System.Int32' cannot be converted to type 'System.Nullable`1[System.Double]'.
   at System.RuntimeType.TryChangeType(Object value, Binder binder, CultureInfo culture, Boolean needsSpecialCast)
   at System.Reflection.MethodBase.CheckArguments(Object[] parameters, Binder binder, BindingFlags invokeAttr, CultureInfo culture, Signature sig)
   at System.Reflection.RuntimeMethodInfo.InvokeArgumentsCheck(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Xunit.Sdk.TestInvoker`1.CallTestMethod(Object testClassInstance) in /_/src/xunit.execution/Sdk/Frameworks/Runners/TestInvoker.cs:line 150
   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_0.<<InvokeTestMethodAsync>b__1>d.MoveNext() in /_/src/xunit.execution/Sdk/Frameworks/Runners/TestInvoker.cs:line 257
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Xunit.Sdk.ExecutionTimer.<AggregateAsync>d__4.MoveNext() in /_/src/xunit.execution/Sdk/Frameworks/ExecutionTimer.cs:line 48
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Xunit.Sdk.ExceptionAggregator.<RunAsync>d__9.MoveNext() in /_/src/xunit.core/Sdk/ExceptionAggregator.cs:line 90
